transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,493 @@
+    /// The channel used to receive new connections.
+    receive_connections: Mutex<Option<mpsc::UnboundedReceiver<ListenerResult>>>,
+    /// Connections send their events to this
+    event_channel: mpsc::Sender<EndpointMessage>,
+    /// The `Multiaddr`
+    address: Multiaddr,
+    /// The configuration
+    config: Config,
+}
+
+impl EndpointData {
+    pub(super) fn event_channel(&self) -> mpsc::Sender<EndpointMessage> {
+        self.event_channel.clone()
+    }
+
+    pub(super) fn socket(&self) -> &socket::Socket {
+        &self.socket
+    }
+}
+
+mod endpoint_ref {
+    use super::{EndpointData, EndpointMessage};
+    use std::sync::Arc;
+    #[derive(Debug, Clone)]
+    pub(super) struct EndpointRef(Option<Arc<EndpointData>>);
+    impl EndpointRef {
+        pub fn new(data: Arc<EndpointData>) -> Self {
+            Self(Some(data))
+        }
+
+        pub fn inner(&self) -> &Arc<EndpointData> {
+            self.0
+                .as_ref()
+                .expect("this is only set to None in the destructor")
+        }
+    }
+
+    impl Drop for EndpointRef {
+        fn drop(&mut self) {
+            let inner = self.0.take().expect("we set this to a Some; qed");
+            let mut drop_channel = inner.event_channel.clone();
+            // decrement the refcount
+            drop(inner);
+            // wake the driver
+            let _ = drop_channel.start_send(EndpointMessage::Dummy);
+        }
+    }
+}
+
+use endpoint_ref::EndpointRef;
+
+/// A QUIC endpoint.  Each endpoint has its own configuration and listening socket.
+///
+/// You generally need only one of these per process.  Endpoints are `Send` and `Sync`, so you
+/// can share them among as many threads as you like.  However, performance may be better if you
+/// have one per CPU core, as this reduces lock contention.  Most applications will not need to
+/// worry about this.  `QuicEndpoint` tries to use fine-grained locking to reduce the overhead.
+///
+/// `QuicEndpoint` wraps the underlying data structure in an `Arc`, so cloning it just bumps the
+/// reference count.  All state is shared between the clones.  For example, you can pass different
+/// clones to `listen_on`.  Each incoming connection will be received by exactly one of them.
+///
+/// The **only** valid `Multiaddr` to pass to `listen_on` or `dial` is the one used to create the
+/// `QuicEndpoint`.  You can obtain this via the `addr` method.  If you pass a different one, you
+/// will get `TransportError::MultiaddrNotSuppported`.
+#[derive(Debug)]
+pub struct Endpoint {
+    data: EndpointRef,
+    join_handle: async_std::task::JoinHandle<Result<(), Error>>,
+}
+
+struct EndpointDriver(Option<Arc<EndpointData>>);
+
+impl Endpoint {
+    /// Construct a `Endpoint` with the given `Config` and `Multiaddr`.
+    pub fn new(
+        config: Config,
+        address: Multiaddr,
+    ) -> Result<Self, TransportError<<&'static Self as Transport>::Error>> {
```

Comment:
This might be controversial, but instead of having `Endpoint` and `EndpointRef` and `EndpointData`, instead I would just return an `Arc<Endpoint>`, use it everywhere, and remove `EndpointRef`/`EndpointData` altogether. This means that we have to remove the `Endpoint::close()` method, but I think that's fine.

Reply:
Sadly, this leads to a task leak: when the last user-held `Arc<Endpoint>` is dropped, the endpoint driver won’t be notified, and may stay blocked forever if nothing else wakes it up.  Sending `EndpointMessage::Dummy` avoids the leak by ensuring that the endpoint driver is woken up.

I can, however, combine `Endpoint` and `EndpointRef`, and make `Endpoint` implement `Clone`.

Reply:
What's wrong with doing `impl Drop for Endpoint {` and sending `EndpointMessage::Dummy` there?

Reply:
Because `EndpointMessage::Dummy` must be sent after the `Arc` is dropped.  The driver task checks the `Arc`’s reference count to determine when it should exit.  If nobody else has a reference to the `quinn_proto::Endpoint`, then there is no way that more activity can happen on the endpoint, so it is okay to exit.  For that to work, it must be awoken after the reference count has been decremented.

Reply:
Well, the `Endpoint` is dropped only after the reference count is at 0, otherwise it would be possible for another thread to access the `Endpoint` being destroyed.

Reply:
The driver task checks the reference count explicitly, using `Arc::strong_count`.  If this is 1, it means that the driver task has the only reference to the `EndpointData`.  That means that there are no connections using the `EndpointData`, and that no new ones can be created.  Hence, the driver task can exit.  This will drop the last reference to the `EndpointData`, which will itself be destroyed.

Reply:
What you're doing looks like what `std::sync::Weak` is designed for

Reply:
I am not certain, but there may be error conditions in which the endpoint driver needs to send off a few more packets before exiting, which requires that the endpoint stay alive.  So `std::sync::Weak` will not work.  In any case, `std::sync::Weak` does not solve the destructor ordering problem.